### ISSUE TITLE
Missing ending slashes for standalone tags

### ISSUE THREAD

**Issue by al-Iskandar** [mblaschke/TYPO3-metaseo#68]
Metaseo (development branch) always uses standalone tags without ending slash although a xhtml version is declared in the template like:

<meta property="og:type" content="article"> 

This breaks the XHTML rules and the compatibility advices.

PS: I think it would be better to make for such functions not commonly used by default the option 0. It is easy for everybody to activate functions he need, but many people who use it not have now more meta tags not effectivily used

**Comment by thomaszbz** [comment]
Which extension do you use for speaking urls? Which version?

Which version of metaseo did you use? A lot has changed in the current development version (which will be the upcoming 2.0.0).

As of my part, using realurl, metaseo (develop) pretty much handels trailing slashes the way realurl does it.

**Comment by al-Iskandar** [comment]
The branch develop I downloaded and copied into the metaseo directory after deleting the repository version (the extension manager showed v2.0.0).

This version breaks also the scheduler in the backend by a class problem, which I could only solve by deleting manually the records for metaseo tasks in the database. It was anything with "Metaseo\Metaseo\Scheduler\Task\GarbageCollectionTask" called by the scheduler.php (ca. line 106) But I do not find anymore the error message and I am happy that my installtion is working again.

**Comment by thomaszbz** [comment]
Please open a separate issue for the GarbageCollectionTask problem (does not correspond to the title of this issue).

Could you please post a minimal xhtml containing all the metaseo generated tags and the list of corresponding validation errors?

**Comment by al-Iskandar** [comment]
The file:

<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html> 
<html xmlns="http://www.w3.org/1999/xhtml" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:sch="http://schema.org/" xml:lang="de">
	<head>
		<meta charset="utf-8" />
		<title>Any Page</title>
		
		
		<meta name="robots" content="index,follow">
		<link rel="canonical" href="http://www.example.com/">
		<meta property="og:site_name" content="Example">
		<meta property="og:title" content="Example">
		<meta property="og:type" content="article">
		
		
		
	</head>
	<body>
		<h1>Any Page</h1>
		<p>Any content.</p>
	</body>
</html>





Message of a browser (SRWare Iron):

This page contains the following errors:
error on line 18 at column 9: Opening and ending tag mismatch: meta line 0 and head
Below is a rendering of the page up to the first error.




The Validator (https://html5.validator.nu/) tells:

    Error: XHTML element link not allowed as child of XHTML element meta in this context. (Suppressing further errors from this subtree.)

    From line 10, column 3; to line 10, column 55

    ollow">↩		<link rel="canonical" href="http://www.example.com/">↩

    Contexts in which element link may be used:
        Where metadata content is expected.
        In a noscript element that is a child of a head element.
        If the itemprop attribute is present: where phrasing content is expected.
    Content model for element meta:
        Empty.
        No end tag.

    Fatal Error: required character (found h) (expected m)

    At line 17, column 4

    d -->↩		↩		↩	</head>↩	<body>↩	


The problem are the missing slashes of the link and meta tags. A second problem would be (after fixing the standalone tags) would be the introduction of the og namespace without declaration (like dc and sch in this example.

**Comment by thomaszbz** [comment]
We need to hear what @mblaschke thinks about it. I'm not so much in the history of the featureset of metaseo. Personally I think that if XHTML is supported, it also needs to validate. The tag generation is one thing, but it's not that easy to properly determine the doctype.

I also filed an issue to Typo3 core:
https://forge.typo3.org/issues/67686

And I filed #70 to cleaner determine the doctype from Typo3 configuration variables

I appreciate it if you come up with a better idea.

**Comment by thomaszbz** [comment]
With the XHTML code you provided I set doctype "XHTML 1.0 strict" in https://validator.w3.org/check and the result is:

    Line 5, Column 23: there is no attribute "charset"
    <meta charset="utf-8" />

    Error Line 5, Column 32: required attribute "content" not specified
    <meta charset="utf-8" />

    Error Line 9, Column 56: end tag for "meta" omitted, but OMITTAG NO was specified
    <meta name="robots" content="index,follow">

    Info Line 9, Column 13: start tag was here
    <meta name="robots" content="index,follow">

    Error Line 10, Column 66: end tag for "link" omitted, but OMITTAG NO was specified
    <link rel="canonical" href="http://www.example.com/">

    Info Line 10, Column 13: start tag was here
    <link rel="canonical" href="http://www.example.com/">

    Error Line 11, Column 28: there is no attribute "property"
    <meta property="og:site_name" content="Example">

    Error Line 11, Column 61: end tag for "meta" omitted, but OMITTAG NO was specified
    <meta property="og:site_name" content="Example">

    Info Line 11, Column 13: start tag was here
    <meta property="og:site_name" content="Example">

    Error Line 12, Column 57: end tag for "meta" omitted, but OMITTAG NO was specified
    <meta property="og:title" content="Example">

    Info Line 12, Column 13: start tag was here
    <meta property="og:title" content="Example">

    Error Line 13, Column 56: end tag for "meta" omitted, but OMITTAG NO was specified
    <meta property="og:type" content="article">

    Info Line 13, Column 13: start tag was here
    <meta property="og:type" content="article">

**Comment by thomaszbz** [comment]
Adding the slashes still returns validation errors:

    <?xml version="1.0" encoding="utf-8"?>
    <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"> 
    <html xmlns="http://www.w3.org/1999/xhtml" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:sch="http://schema.org/" xml:lang="de">
        <head>
            
            <title>Any Page</title>
    
            
                <meta name="robots" content="index,follow" />
                <link rel="canonical" href="http://www.example.com/" />
                <meta property="og:site_name" content="Example" />
                <meta property="og:title" content="Example" />
                <meta property="og:type" content="article" />
            
    
        </head>
        <body>
            <h1>Any Page</h1>
            <p>Any content.</p>
        </body>
    </html>

Result:

    Line 11, Column 28: there is no attribute "property"
    <meta property="og:site_name" content="Example" />

**Comment by thomaszbz** [comment]
There's a similar issue on forge: https://forge.typo3.org/issues/62154

**Comment by al-Iskandar** [comment]
It should be XML-HTML 5. But I did not copy all my page which validates with following message (https://html5.validator.nu/) "The document is valid XHTML5 + ARIA + SVG 1.1 + MathML 2.0 (subject to the utter previewness of this service)."

Also I mentioned before, that I expect a namespace error when using endslashes, because og was a new feature and not implemented in the old system I used, because it had namespaces for XHTML, dc terms and schema.org. I do not know whether you can get a valid page with XHTML 1.0 Strict. I think you need minimum XHTML+RDFa 1.x or XHTML 5.

The problem of Typo3 Core is that it has no direct way for XHTML5. You must declare that it is HTML5 with XHTML specifications. But 

    ################
    # HTML-Ausgabe #
    ################

    # XHTML-Anpassungen:
    config.xmlprologue = <?xml version="1.0" encoding="utf-8"?>
    config.htmlTag_setParams = xmlns="http://www.w3.org/1999/xhtml" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:sch="http://schema.org/" xml:lang="de"

    # Festlegung des HTML-Standards:
    config.doctype = html5

    # Festlegung des XHTML-Standards:
    config.xhtmlDoctype = xhtml_11

    # Elementare Überarbeitung auf XHTML-Grundregeln:
    # config.xhtml_cleaning = output (now deprecated - till 6.2 the solution for this problem)

    # Anpassung des MIME-Types auf application/xhtml+xml:
    config.additionalHeaders = content-type: application/xhtml+xml;charset=utf-8

    # Zeichenkodierung festlegen:
    # config.renderCharset = utf-8

**Comment by thomaszbz** [comment]
To be honest, I did not hear a lot about XHTML5. I was a bit surprised that there's now also a XHTML5 in the specs (and in a way I feel happy with that...). Initially I thought you wanted to go for the common XHTML 1.x. 

I guess your TYPO3 config is mismatch. Here's some options:

- When you set config.doctype=html5 we think you want "HTML5".
- There is no xhmtl5 constant to be used yet, neither for doctype nor for xhtmlDoctype (to my knowledge)
- config.xhtmlDoctype MUST be set if config.doctype contains a custom doctype string. Using a doctype string would be an option, but xhtml5 is not defined for xhtmlDoctype
- Another approach can be found here: https://forge.typo3.org/issues/40503

For the moment, I think that XHTML5 is not supported by TYPO3. You could set a custom string in config.doctype. But then there's no newer xhtmlDoctype for XHTML5 available (not sure if that's really bad).

When you set a custom string and a bad value for xhtmlDoctype, the problem is: How should an extension determine that you want to do XHTML5?

I filed two issues, already mentioning XHTML5:
#70 and
https://forge.typo3.org/issues/67686

I guess, in the first place, you need to come up with a clean TYPO3 configuration which supports XHTML5. In my view, this is not supported yet, because there is no xhtmlDoctype xhtml_5 yet.

There is a ticket for XHTML5 support in TYPO3: https://forge.typo3.org/issues/40503

A workaround could be that we introduce a new extension specific configuration variable for the doctype. Then this would mean that all extensions needed to do that to render doctype specific content. That does not really make sense to me, but if there should be a metaseo-specific-workaround temporarily needs to be decided by @mblaschke .

By the way: Do major browser implementations support XHTML5 already?

**Comment by al-Iskandar** [comment]
I am running two pages with valid XHTML 5 and a third is shortly before the launch. They run with Firefox, Opera (The last old style version), Chromium and IE as long as they support HTML5.
For avoiding compatibility hacks I avoid new elements I do not need necessarly, also I do not use elements which are for me a backway to HTML stoneage (e.g. the zombie embed).
I prefer XHTML because the code is cleaner and more logical, also I like the namespace concept.
The configuration is not the optimum, but so I get a HTML of the 5th generation which is XML based.

If you understand German: http://webkompetenz.wikidot.com/html-handbuch:xhtml-dokumenttypen / http://webkompetenz.wikidot.com/html-handbuch:xhtml-unterschiede from the online version of the German reference for HTML5. (The doctype declaration is explicitly allowed in XHTML 5 although not necessary.

**Comment by al-Iskandar** [comment]
PS: HTML 5 and XHTML 5 are very similar. XHTML 5 must be according to XML rules, use the MIME type application/xhtml+xml and use namespaces.

**Comment by thomaszbz** [comment]
Before HTML 5 came up, I preferred XHTML over HTML for the same reasons.

Now the issue here is not a matter of taste but XHTML 5 support by TYPO3 and the question how an extension can determine the doctype.

Metaseo needs to know if XHTML 5 shall be used because we did a lot of changes to validate against HTML5 (which is very similar to XHTML5, as you say). It's not only the syntax, but also supported meta tags, etc.

**Comment by al-Iskandar** [comment]
I am not a programmer, but I think a signal could be the use of config.xhtmlDoctype that just some additions like the backslash are made. Alternatively a solution would be, to use ending slashes also in SGML-HTML 5 which is, as far I know, now allowed in SGML-HTML 5.

http://www.w3.org/TR/html-polyglot/

**Comment by thomaszbz** [comment]
Hm, that's misuse of configuration variables which would be obsolete as soon as TYPO3 comes up with XHTML 5. Then there's breaking changes for users using misused configuriation variables and all that jazz.

Plus, if TYPO3 is not supporting XHTML 5 it's not working in other extensions either. Therefore it seems a bit masochistic to me at the moment.

Personally, I changed from XHTML 1.x to HTML5. Maybe I go for XHTML 5 again as soon as it's supported and widely used (community and browsers).

You are right, adding a few slashes sounds easy. We'll see what @mblaschke thinks about it. He knows WAYS more about the history of metaseo and supported doctypes than me.

**Comment by thomaszbz** [comment]
There's multiple issues in class MetatagPart:

    public function main() {
        [...]
        $this->metaTagList = $this->renderMetaTags($this->metaTagList);
        return $this->metaTagList;
    }

means that a type change happens in $this->metaTagList from array to string, because renderMetaTags returns a string (including line separators).

Plus, all the slash magic happens in renderMetaTags():

    protected function renderMetaTags(array $metaTags) {
        $ret = array();
        $isHtml5 = $this->isHtml5();
        foreach ($metaTags as $metaTag) {
            $tag        = $metaTag['tag'];
            $attributes = array();
            foreach ($metaTag['attributes'] as $key => $value) {
                $attributes[] = $key . '="' . htmlspecialchars($value) . '"';
            }
            if ($isHtml5) {
                $ret[] = '<' . $tag . ' ' . implode(' ', $attributes) . '>';
            } else {
                $ret[] = '<' . $tag . ' ' . implode(' ', $attributes) . '/>';
            }
        }
        $separator = "\n";
        return $separator . implode($separator, $ret) . $separator;
    }

Basically that means that you should always get slashes unless you use config.doctype="html5". Easy solution would be to also require that xhtmlDoctype is not set. Then you'd get html5 with slashes ;)

Another option could be to use

    $GLOBALS['TSFE']->getPageRenderer()->addMetaTag()

in renderMetaTags() instead and see what happens.

**Comment by thomaszbz** [comment]
In \TYPO3\CMS\Frontend\Page\PageGenerator, they do it that way:

    switch ((string) $GLOBALS['TSFE']->config['config']['xmlprologue']) {
        case 'none':
            $xmlDocument = FALSE;
            break;

        case 'xml_10':
        case 'xml_11':
        case '':
        default:
    }

And then TYPO3 processes all the meta tags and removes slashes for html5 when XML is not to be used:

    case 'html5':
        $docTypeParts[] = '<!DOCTYPE html>';
        if ($xmlDocument) {
            $pageRenderer->setMetaCharsetTag('<meta charset="|" />');
        } else {
            $pageRenderer->setMetaCharsetTag('<meta charset="|">');
        }
        break;

and

    $metaTagsHtml = static::generateMetaTagHtml(
            isset($GLOBALS['TSFE']->pSetup['meta.']) ? $GLOBALS['TSFE']->pSetup['meta.'] : array(),
            $GLOBALS['TSFE']->xhtmlVersion,
            $GLOBALS['TSFE']->cObj
        );


    static protected function generateMetaTagHtml(array $metaTagTypoScript, $xhtml, ContentObjectRenderer $cObj) {
        // Add ending slash only to documents rendered as xhtml
        $endingSlash = $xhtml ? ' /' : '';

I think we should be compliant with the core. Therefore we need to detect XHTML and not set slashes here. Or just use

    $GLOBALS['TSFE']->getPageRenderer()->addMetaTag()

which basically does the same and is maintained by the TYPO3 community.

**Comment by mblaschke** [comment]
Can you please check if the last commit is working for this issue?

**Comment by thomaszbz** [comment]
Your configuration returns 110 for $GLOBALS['TSFE']->xhtmlVersion using TYPO3 6.2

**Comment by thomaszbz** [comment]
oh, you were faster, @mblaschke  ;)

I used the

    $GLOBALS['TSFE']->xhtmlVersion

approach TYPO3 uses. Plus, my version only removes slashes when html5 and no xml is set.

I think your version is removing slashes also for HTML!=5.

In the end I don't mind - if it validates.

**Comment by al-Iskandar** [comment]
A first check toold me that browser and validator accept the new version.
The users needs only to add the correct namespace: xmlns:og="http://ogp.me/ns#", but who uses XML parsed XHTML should know that and add it manually or deactivate the open graph part according to his needs.
So it seems to be solved for the side of metaseo - Typo3 Core and XHTML5 is another story without end (I know issue Feature #40503 since long long time - the last three days it got suddenly more life than in the three years before)

**Comment by thomaszbz** [comment]
User is happy, issue is closed. We'll come back to this topic as soon as TYPO3 officially offers XHTML5 support.
